File: app/src/main/java/info/nightscout/androidaps/plugins/pump/medtronic/comm/ui/MedtronicUIComm.java
package info.nightscout.androidaps.plugins.pump.medtronic.comm.ui;

import javax.inject.Inject;

import dagger.android.HasAndroidInjector;
import info.nightscout.androidaps.logging.AAPSLogger;
import info.nightscout.androidaps.logging.LTag;
import info.nightscout.androidaps.plugins.pump.medtronic.comm.MedtronicCommunicationManager;
import info.nightscout.androidaps.plugins.pump.medtronic.defs.MedtronicCommandType;
import info.nightscout.androidaps.plugins.pump.medtronic.util.MedtronicUtil;

/**
 * Created by andy on 6/14/18.
 */
public class MedtronicUIComm {

    private final HasAndroidInjector injector;
    private final AAPSLogger aapsLogger;
    private final MedtronicUtil medtronicUtil;
    private final MedtronicCommunicationManager medtronicCommunicationManager;
    private final MedtronicUIPostprocessor medtronicUIPostprocessor;

    @Inject
    public MedtronicUIComm(
            HasAndroidInjector injector,
            AAPSLogger aapsLogger,
            MedtronicUtil medtronicUtil,
            MedtronicUIPostprocessor medtronicUIPostprocessor,
            MedtronicCommunicationManager medtronicCommunicationManager
    ) {
        this.injector = injector;
        this.aapsLogger = aapsLogger;
        this.medtronicUtil = medtronicUtil;
        this.medtronicUIPostprocessor = medtronicUIPostprocessor;
        this.medtronicCommunicationManager = medtronicCommunicationManager;
    }

    public synchronized MedtronicUITask executeCommand(MedtronicCommandType commandType, Object... parameters) {

        aapsLogger.warn(LTag.PUMP, "Execute Command: " + commandType.name());

        MedtronicUITask task = new MedtronicUITask(injector, commandType, parameters);

        medtronicUtil.setCurrentCommand(commandType);

        // new Thread(() -> {
        // LOG.warn("@@@ Start Thread");
        //
        // task.execute(getCommunicationManager());
        //
        // LOG.warn("@@@ End Thread");
        // });

        task.execute(medtronicCommunicationManager);

        // for (int i = 0; i < getMaxWaitTime(commandType); i++) {
        // synchronized (task) {
        // // try {
        // //
        // // //task.wait(1000);
        // // } catch (InterruptedException e) {
        // // LOG.error("executeCommand InterruptedException", e);
        // // }
        //
        //
        // SystemClock.sleep(1000);
        // }
        //
        // if (task.isReceived()) {
        // break;
        // }
        // }

        if (!task.isReceived()) {
            aapsLogger.warn(LTag.PUMP, "Reply not received for " + commandType);
        }

        task.postProcess(medtronicUIPostprocessor);

        return task;

    }

    public int getInvalidResponsesCount() {
        return medtronicCommunicationManager.getNotConnectedCount();
    }
}

File: pump/medtronic/src/main/kotlin/app/aaps/pump/medtronic/comm/ui/MedtronicUIComm.kt
package app.aaps.pump.medtronic.comm.ui

import app.aaps.core.interfaces.logging.AAPSLogger
import app.aaps.core.interfaces.logging.LTag
import app.aaps.pump.medtronic.comm.MedtronicCommunicationManager
import app.aaps.pump.medtronic.defs.MedtronicCommandType
import app.aaps.pump.medtronic.util.MedtronicUtil
import javax.inject.Inject
import javax.inject.Provider

/**
 * Created by andy on 6/14/18.
 */
class MedtronicUIComm @Inject constructor(
    private val aapsLogger: AAPSLogger,
    private val medtronicUtil: MedtronicUtil,
    private val medtronicUIPostprocessor: MedtronicUIPostprocessor,
    private val medtronicCommunicationManager: MedtronicCommunicationManager,
    private val medtronicUITaskProvider: Provider<MedtronicUITask>
) {

    fun executeCommand(commandType: MedtronicCommandType): MedtronicUITask {
        return executeCommand(commandType, null)
    }

    @Synchronized
    fun executeCommand(commandType: MedtronicCommandType, parameters: ArrayList<Any>?): MedtronicUITask {

        aapsLogger.info(LTag.PUMP, "Execute Command: " + commandType.name)
        val task = medtronicUITaskProvider.get().with(commandType, parameters)
        medtronicUtil.setCurrentCommand(commandType)
        task.execute(medtronicCommunicationManager)

        if (!task.isReceived) {
            aapsLogger.warn(LTag.PUMP, "Reply not received for $commandType")
        }

        task.postProcess(medtronicUIPostprocessor)
        return task
    }

    val invalidResponsesCount: Int
        get() = medtronicCommunicationManager.getNotConnectedCount()

}